### PR TITLE
Fix race in PageProcessor

### PR DESCRIPTION
**Pull request by cberner**
Fix race which could cause queries to fail when using
concat(array, array), or when enabling columnar processing

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java [hunk 1]
```diff
@@ -82,23 +82,29 @@ public long getCacheSize()
 
     public CursorProcessor compileCursorProcessor(RowExpression filter, List<RowExpression> projections, Object uniqueKey)
     {
-        return cursorProcessors.getUnchecked(new CacheKey(filter, projections, uniqueKey));
+        try {
+            return cursorProcessors.getUnchecked(new CacheKey(filter, projections, uniqueKey)).newInstance();
```
Comment (nileema) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java]: put `newInstance()` on a new line

File: presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java [hunk 2]
```diff
@@ -82,23 +82,29 @@ public long getCacheSize()
...
+        catch (ReflectiveOperationException e) {
+            throw Throwables.propagate(e);
+        }
     }
 
     public PageProcessor compilePageProcessor(RowExpression filter, List<RowExpression> projections)
     {
-        return pageProcessors.getUnchecked(new CacheKey(filter, projections, null));
+        try {
+            return pageProcessors.getUnchecked(new CacheKey(filter, projections, null)).newInstance();
```
Comment (nileema) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java]: same

**Comment by nileema** [comment]
Assuming tests pass, looks good.

**Comment by martint** [comment]
This needs a better commit message. Explain that it's really an issue with instances of PageProcessors being shared across threads (which were previously assumed to be stateless) due to how caching works and that we're changing to caching the generated classes instead.

**Comment by martint** [comment]
Also, update the release notes of the versions affected to warn about the broken behavior

**Pull request merged**